File: leetcode0684/solution.go
/*
Solution: union find

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 5.00 MB | 42.03%

- n: length of nodes
*/

func findRedundantConnection(edges [][]int) []int {
	n := len(edges)

	var parent []int
	var degree []int

	for i := range n + 1 {
		parent = append(parent, i)
		degree = append(degree, 1)
	}

	var res []int

	for _, edge := range edges {
		u := edge[0]
		v := edge[1]

		if !union(u, v, parent, degree) {
			res = append(res, u)
			res = append(res, v)

			return res
		}
	}

	return res
}

func find(node int, parent []int) int {
	p := parent[node]

	for p != parent[p] {
		p = parent[p]
	}

	return p
}

func union(u int, v int, parent []int, degree []int) bool {
	parent_u := find(u, parent)
	parent_v := find(v, parent)

	if parent_u == parent_v {
		return false
	}

	if degree[parent_u] > degree[parent_v] {
		parent[parent_v] = parent_u
		degree[parent_u] += degree[parent_v]
	} else {
		parent[parent_u] = parent_v
		degree[parent_v] += degree[parent_u]
	}

	return true
}

File: leetcode0684/solution.rs
/*
Solution: union find

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 2.82 MB | 66.67%

- n: length of nodes
*/

impl Solution {
    pub fn find_redundant_connection(edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n: usize = edges.len();
        let mut roots: Vec<usize> = (0..=n).map(|idx| idx).collect::<Vec<usize>>();
        let mut degree: Vec<i32> = vec![1; n + 1];

        let mut redundant: Vec<Vec<i32>> = Vec::new();

        for edge in edges.iter() {
            let u: usize = edge[0] as usize;
            let v: usize = edge[1] as usize;

            if roots[u] == roots[v] {
                redundant.push(vec![u as i32, v as i32]);
            } else {
                if degree[roots[u]] >= degree[roots[v]] {
                    let root: usize = roots[v];
                    for i in 0..=n {
                        if roots[i] == root {
                            degree[roots[i]] -= 1;
                            roots[i] = roots[u];
                            degree[roots[i]] += 1;
                        }
                    }
                } else {
                    let root: usize = roots[u];
                    for i in 0..=n {
                        if roots[i] == root {
                            degree[roots[i]] -= 1;
                            roots[i] = roots[v];
                            degree[roots[i]] -= 1;
                        }
                    }

                }
            }
        }

        redundant.pop().unwrap()
    }
}